neutral_generation/smart_convert.py

```diff
@@ -0,0 +1,284 @@
+
+    # identify all verbs
+    verbs = set()
+    # this deals with repeating verbs, e.g. "He sings and sings."
+    #   because verb with same text will have different position (makes them unique)
+    for possible_subject in doc:
+        is_subject = (
+            possible_subject.dep == nsubj and                   # current token is a subject
+            # head of current token is a verb
+            (possible_subject.head.pos == VERB or possible_subject.head.pos == AUX) and
+            possible_subject.text.lower() in SUBJECT_PRONOUNS   # current token is either she / he
+        )
+        if is_subject:
+            verbs.add(possible_subject.head)
+
+    # identify all conjuncts and add them to set of verbs
+    # e.g. he dances and prances --> prances would be a conjunct
+    # TODO: understand all English labels of dependency parse better
+    for possible_conjunct in doc:
+        is_conjunct = (
+            possible_conjunct.dep == conj and           # current token is a conjunct
+            # removing the line below, can have a conjunct with an auxiliary verb, e.g. "He was angry and ran away."
+            # possible_conjunct.head.pos == VERB and      # head of current token is a verb
+            possible_conjunct.head in verbs             # the subject of that verb is she / he
+        )
+        if is_conjunct:
+            verbs.add(possible_conjunct)
+
+    # print('verbs: ', verbs)
+
+    auxiliaries = dict()
+    for verb in verbs:
+        auxiliaries[verb] = list()
+    for possible_aux in doc:
+        is_auxiliary = (
+            # current token is an auxiliary verb or negation (not)
+            (possible_aux.dep == aux or possible_aux.dep == neg) and
+            possible_aux.head.pos == VERB and       # head of current token is a verb
+            possible_aux.head in verbs              # the subject of that verb is she / he
+        )
+        if is_auxiliary:
+            verb = possible_aux.head
+            auxiliaries[verb].append(possible_aux)
+
+    print('verbs: ', auxiliaries)
+
+    return auxiliaries
+
+
+def pluralize_verbs(s, auxiliaries):
+    """
+    :param s: input sentence
+    :param auxiliaries: dictionary of verbs and list of auxiliaries (if any) to be pluralized
+    :return: sentence with plural form of verbs
+    """
+
+    # do_not_change includes 7 tenses in total:
+    #   the 4 future tenses, the 2 past perfect tenses, and the past simple
+    do_not_change = ['past simple', 'past perfect', 'future']
+
+    # rule_change includes 4 tenses in total:
+    #   the present continuous, the 2 present perfect tenses, and the past continuous
+    rule_change = ['present continuous', 'present perfect', 'past continuous']
+
+    # TODO: categorize tense for "do" verbs
+    verb_tense = dict()
+    for verb, auxiliary_list in auxiliaries.items():
+        auxiliary_text = ' '.join(auxiliary.text for auxiliary in auxiliary_list)
+        full_verb = auxiliary_text + ' ' + verb.text
```

Comment:
String concat using '+' in python is inefficient. If you want to optimize your code, consider alternatives.
See: https://blog.mclemon.io/python-efficient-string-concatenation-in-python-2016-edition

Reply:
Very interesting! Thanks for sharing this tidbit. I'll replace string concatenation with ''.join(...) from now on.